### DISCUSSION TITLE
How to add new libs to our FFmpeg build

### DISCUSSION THREAD

**Discussion by lukasf** [ffmpeginteropx/FFmpegInteropX#261] [General]
Adding new libs is not trivial. This discussion should help you on your journey to add new libs.

## Compile for UWP using MSVC ##

The first step is to get the actual lib compiled for UWP using the MSVC toolchain, either via direct Windows call, or through our MSYS2 linux emulation layer. This is the most difficult part: Find out how the lib's build system works. Check if and how it is possible to change the compiler. Change it and hope it works. Check if there is a UWP configuration, create one if it is not there. Make sure if builds on all target platforms. There is no general advice here. You need to dig in and find a way. It's not a bad idea to ask the lib developers. If you are lucky, they use github or similar systems, where you could open an issue. Some use mailing lists, which make it much harder (at least for me) to get in touch. Explain them that you need to build with MSVC because the lib needs to be linked with MSVC based project.

Most libs come from a linux environment and use gcc as compiler. There is a vast number of different build systems on linux. I know a few by now, but there are a whole lot more. Some make it easy to switch compilers, for some it is very difficult.

Sometimes you can find forks that have been changed to use MSVC, but often they are outdated. The [ShiftMediaProject](https://github.com/ShiftMediaProject) has a lot of libs for FFmpeg, but not all of them. I forked a bunch of them and added ARM/ARM64 targets. But for many others, I had to manually find out how to build them with MSVC so they properly link with our FFmpeg build.

A few tips:

Often, linux scripts will use `CC` (the GCC compiler) for compiler calls. You might need to replace it with MSVC cl.exe like this `CC=cl`. Either this can be done as some parameter, or you need to put this right before the build or configure call with only a space separator, e.g. `CC=cl ./configure`.

You might have to manually create an UWP build configuration. These are the parameters required (you might have to change `/` to `-`.
Compiler arguments (c_args): /MD /DWINAPI_FAMILY=WINAPI_FAMILY_APP
Linker arguments (c_link_args): /APPCONTAINER /WINMD WindowsApp.lib

## How the FFmpeg configure script works ##

When you enable a new lib in ffmpeg, the confgure script will check that the lib is actually found with the expected name, that the include files are found in the right location, and that compilation will work. This is usually done with calls like this in the configure script.

```
enabled  zlib && { check_pkg_config zlib zlib "zlib.h" zlibVersion ||
                   check_lib zlib   zlib.h      zlibVersion    -lz; }
enabled bzlib && check_lib bzlib bzlib.h BZ2_bzlibVersion    -lbz2
enabled  lzma && check_lib lzma   lzma.h lzma_version_number -llzma
```

So what happens here?

- Check pkg-config to see if lib is listed (optional, not done for all libs)
  - We do not use pkg-config and it is not possible to disable this check. So I created a projekt "PkgConfigFake" that will just return what the configure script expects. If you find an error in the configure.log that the lib was not found using pkg-config, you need to change the PkgConfigFake so it will return proper values for your lib. Usually the configure script just checks for existance, sometimes it even checks the version number. Just fake it all.
- Check if the lib file is found
  - The difficult part here is finding out what name ffmpeg expects. Often it is the name that it used internally, sometimes it is the last parameter of the configure call (-lbz2 means "link bz2.lib"). Sometimes it is none of the two and I don't even know how ffmpeg gets the name.
- Check if the header files are found (at least one of them is usually checked)
  - They need to be in the right folder. For some libs, they just need to be in "inc" folder, for some it is a subfolder. For almost all libs, the set of include files required here is generated during build or install step of the lib. You usually can't just copy the .h files from the lib's source dir. You need to find out where the inc files are put during build/install and copy them to our ffmpeg intermediate directory.
- Try to compile a small .exe file with one very basic lib call
  - This makes sure that the lib will actually build and link with the parameters used for ffmpeg build. Sometimes you need to add lib specific perprocessor definitions to the ffmpeg build, to make the build succeed (e.g. zlib need this).

## Integrate with our FFmpeg build script ##

If you have succeeded to (manually) build the lib and link it with ffmpeg, the most difficult parts are done. Now you need to automate the steps and integrate them in the build script. Make sure that build works for all platforms. There might be problems with assembler code for specific platforms, some errors only show up late when linking ffmpeg.

**Comment by brabebhin** [comment]
Thanks for this guide. This makes things a bit clearer, especially the seemingly random strings in the fake package config.
